### PR TITLE
Improve format of configuration file syntax errors

### PR DESCRIPTION
**Pull request by adityadivekar03**
Change the format of configuration file syntax errors to make them more
user friendly.

fixes [#3437](http://trac.buildbot.net/ticket/3437)

### CODE REVIEW

File: master/buildbot/config.py [hunk 1]
```diff
@@ -201,8 +202,9 @@ def loadConfig(cls, basedir, filename):
                 raise errors
             except Exception:
                 log.err(failure.Failure(), 'error while parsing config file:')
-                error("error while parsing config file: %s (traceback in logfile)" %
-                      (sys.exc_info()[1],),
+                error("error while parsing config file"
```
Comment (rutsky) [inline master/buildbot/config.py]: space is missed here, message will be like `... file(full ...`

**Comment by adityadivekar03** [comment]
Sorry for the mistake. I've inserted the space now.

**Comment by rutsky** [comment]
Thanks!

LGTM.

File: master/buildbot/config.py [hunk 2]
```diff
@@ -201,8 +202,9 @@ def loadConfig(cls, basedir, filename):
                 raise errors
             except Exception:
                 log.err(failure.Failure(), 'error while parsing config file:')
-                error("error while parsing config file: %s (traceback in logfile)" %
-                      (sys.exc_info()[1],),
+                error("error while parsing config file "
+                      "(full traceback in logfile):\n%s" %
```
Comment (tardyp) [inline master/buildbot/config.py]: I agree parse error should be in the stderr.

but it looks like now the full traceback is not only in the logfile

so I would remove (full traceback in logfile)

**Comment by rutsky** [comment]
> so I would remove (full traceback in logfile)

Then there will be no traceback in the logfile, which is bad IMO, log must be self-sufficient.

Log is duplicated in the user console when user does `buildbot start ./master`, because `start` or `reconfig` commands follows and outputs log too, AFAIK.
But as I saw they not really duplicated because of Python/Twisted issues/misconfiguration:
traceback that goes into logfile looks like doesn't have frames from inside `eval` session, and traceback that is formatted by `traceback.format_exc()` doesn't include frames outside `eval`.

Here is a log that I saw with changes from this PR:

```
2016-02-02 12:02:54+0300 [-] Loading configuration from '/home/bob/stuff/buildbot/master/master.cfg'
2016-02-02 12:02:54+0300 [-] error while parsing config file:
	Traceback (most recent call last):
	  File "/home/bob/stuff/buildbot/env/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 246, in inContext
	    result = inContext.theWork()
	  File "/home/bob/stuff/buildbot/env/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 262, in <lambda>
	    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
	  File "/home/bob/stuff/buildbot/env/local/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "/home/bob/stuff/buildbot/env/local/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
	    return func(*args,**kw)
	--- <exception caught here> ---
	  File "/home/bob/stuff/buildbot/buildbot/master/buildbot/config.py", line 204, in loadConfig
	    exec(f, localDict)
	exceptions.SyntaxError: invalid syntax (master.cfg, line 4)
	
2016-02-02 12:02:54+0300 [-] Configuration Errors:
2016-02-02 12:02:54+0300 [-]   error while parsing config file (full traceback in logfile):
	Traceback (most recent call last):
	  File "/home/bob/stuff/buildbot/buildbot/master/buildbot/config.py", line 204, in loadConfig
	    exec(f, localDict)
	  File "/home/bob/stuff/buildbot/master/master.cfg", line 4
	    c = BuildmasterConfig = {}}
	                              ^
	SyntaxError: invalid syntax
```

@adityadivekar03, can you investigate how to get complete in terms of stack frames and user-friendly looking stack trace?

**Comment by adityadivekar03** [comment]
@rutsky , If I'm not wrong I should look into how to include the frames from inside eval into the logfile so as to make the logfile complete?

**Comment by rutsky** [comment]
> @rutsky , If I'm not wrong I should look into how to include the frames from inside eval into the logfile so as to make the logfile complete?

Yes. This is not obligatory IMO, but if this issue can be easily solved either by adding flags to `traceback.format_exc()` or by not simply reraising exception, but doing something more advanced (e.g. by using Twisted's Failure), that would be nice to have.

**Comment by adityadivekar03** [comment]
Could you tell me how can I check this issue locally, so that I can check my fix locally first?

**Comment by rutsky** [comment]
Sure! You need to create master configuration and add an error to it.
Follow [this tutorial](http://trac.buildbot.net/wiki/RunningBuildbotWithVirtualEnv) until master start part:

```
buildbot create-master master-dir
mv master-dir/master.cfg.sample master-dir/master.cfg 
# add error to master.cfg
echo "XXX" >> master-dir/master.cfg
buildbot start ./master-dir
# take a look at traceback
```

**Comment by adityadivekar03** [comment]
I tried many approaches using the functions from the Failure class from Twisted, but unfortunately couldn't come to a fix. 
I am not able to figure out how to include the line that's being executed, in the logged file.  
Any hints if possible?

**Comment by rutsky** [comment]
I think Twisted simply doesn't handle `SyntaxError` in the same way, as Python does.
Python error handler treats `SyntaxError` in a [special way](https://docs.python.org/2.7/library/traceback.html?highlight=syntaxerror#traceback.print_exception):

>  if etype is SyntaxError and value has the appropriate format, it prints the line where the syntax error occurred with a caret indicating the approximate position of the error.

Lets handle `SyntaxError` in separate `except SyntaxError:` clause and output `traceback.format_exc()` there (and no changes in `except Exception` clause at all).

**Comment by adityadivekar03** [comment]
I think the tests are failing because the tests are configured to assert against a predefined handling involving the earlier Exception clause, but according to the commit the syntax errors are raised separately, and hence the message displayed has changed. 
The log in my system on adding a syntax error in the master.cfg file is as follows -
<pre><code>
2016-02-20 17:17:27+0530 [-] Log opened.
2016-02-20 17:17:27+0530 [-] twistd 15.5.0 (/home/hp/buildbot-work/sandbox/bin/python 2.7.6) starting up.
2016-02-20 17:17:27+0530 [-] reactor class: twisted.internet.epollreactor.EPollReactor.
2016-02-20 17:17:27+0530 [-] Starting BuildMaster -- buildbot.version: 0.9.0b7-8-gc4423de
2016-02-20 17:17:27+0530 [-] Loading configuration from '/home/hp/buildbot-work/sandbox/master-dir/master.cfg'
2016-02-20 17:17:27+0530 [-] Configuration Errors:
2016-02-20 17:17:27+0530 [-]   encountered a SyntaxError while parsing config file:
	Traceback (most recent call last):
	  File "/home/hp/buildbot-work/src/master/buildbot/config.py", line 198, in loadConfig
	    exec(f, localDict)
	  File "/home/hp/buildbot-work/sandbox/master-dir/master.cfg", line 11
	    c = BuildmasterConfig = {}}
	                              ^
	SyntaxError: invalid syntax
	 
2016-02-20 17:17:27+0530 [-] Halting master.
2016-02-20 17:17:27+0530 [-] BuildMaster is stopped
2016-02-20 17:17:27+0530 [-] Main loop terminated.
2016-02-20 17:17:27+0530 [-] Server Shut Down.
</code></pre>

**Comment by rutsky** [comment]
@adityadivekar03 
>  Should I look into how the test cases are to be changed?

Tests can be changed if you reasonably change behavior. So please fix them.

Your log looks great! What do you in the console?

**Comment by adityadivekar03** [comment]
The same log as above is displayed in the terminal. 
There is no duplication of logs as was the case in the earlier code. 
Also I think, the error displayed in terminal is apt now since it only displays those calls to files that the user can verify, i.e. only the concerned frames are displayed now. 
I'll start working on the test cases then. Thanks!

**Comment by adityadivekar03** [comment]
@rutsky the tests are working fine now. Removed the flushLoggedErrors function call since we are not adding the SyntaxError to the list of errors, but simply creating a log event when it is encountered. The list of errors is used only for adding ConfigErrors and hence we do not add the SyntaxError to the list.

**Comment by rutsky** [comment]
Thanks, @adityadivekar03, your changes LGTM!

**Comment by adityadivekar03** [comment]
Thanks a lot for your help!

**Pull request merged**